secretstorage/util.py

```diff
@@ -129,15 +129,17 @@ def exec_prompt(connection: DBusConnection,
 	          is a list of unlocked object paths
 	"""
 	prompt = DBusAddressWrapper(prompt_path, PROMPT_IFACE, connection)
-	dismissed = result = None
-	def callback(msg_body: Tuple[bool, List[str]]) -> None:
-		_dismissed, _result = msg_body
-		nonlocal dismissed, result
-		dismissed, result = bool(_dismissed), _result
-	connection.router.subscribe_signal(callback, prompt_path, PROMPT_IFACE, 'Completed')
-	prompt.call('Prompt', 's', '')
-	if result is None:
-		connection.recv_messages()
+	rule = MatchRule(
+		path=prompt_path,
+		interface=PROMPT_IFACE,
+		member='Completed',
+		type=MessageType.signal,
+	)
+	with connection.filter(rule) as signals:
+		prompt.call('Prompt', 's', '')
+		while len(signals) == 0:
+			connection.recv_messages()
+	dismissed, result = signals.popleft().body
```

Comment:
This is the main functional change. Callbacks are going away in favour of putting messages in queues. It's not very elegant in the blocking integration, because you have to have a loop to pull messages until one goes into the queue, but I think it's still nicer than the callback.